tests/core/test_core_components.py

```diff
@@ -327,9 +338,127 @@ def test_LayerStructure(
             np.equal(template["layer_roles"].to_numpy(), layer_structure.layer_roles)
         )
         assert np.all(
-            np.equal(template["cell_id"].to_numpy(), np.arange(layer_structure.n_cells))
+            np.equal(
+                template["cell_id"].to_numpy(), np.arange(layer_structure._n_cells)
+            )
+        )
+
+
+@pytest.mark.parametrize(
+    argnames="requested,as_bool,raises,expected",
+    argvalues=[
+        pytest.param(
+            "above",
+            True,
+            does_not_raise(),
+            np.array([True] + [False] * 13),
+            id="basic bool",
+        ),
+        pytest.param(
+            "above",
+            False,
+            does_not_raise(),
+            np.array([0]),
+            id="basic int",
+        ),
+        pytest.param(
+            ["above", "subsoil"],
+            True,
+            does_not_raise(),
+            np.array([True] + [False] * 12 + [True]),
+            id="aggregate bool",
+        ),
+        pytest.param(
+            ["above", "subsoil"],
+            False,
+            does_not_raise(),
+            np.array([0, 13]),
+            id="aggregate int",
+        ),
+        pytest.param(
+            "above_canopy",
+            False,
+            pytest.raises(ValueError),
+            None,
+            id="bad string",
+        ),
+        pytest.param(
+            ["above", "lowsoil"],
+            False,
+            pytest.raises(ValueError),
+            None,
+            id="bad list",
+        ),
+    ],
+)
+def test_LayerStructure_get_index(
+    caplog, fixture_core_components, requested, as_bool, raises, expected
+):
+    """Test the get index method."""
+
+    with raises:
+        indices = fixture_core_components.layer_structure.get_indices(
+            requested, as_bool=as_bool
         )
 
+    if isinstance(raises, does_not_raise):
+        assert np.all(np.equal(indices, expected))
+
+
+def test_LayerStructure_set_filled_canopy():
+    """Test the set_filled_canopy_method."""
+
+    from virtual_ecosystem.core.config import Config
+    from virtual_ecosystem.core.core_components import LayerStructure
+
+    cfg = Config(cfg_strings="[core]")
+    layer_structure = LayerStructure(
+        cfg, n_cells=9, max_depth_of_microbial_activity=0.25
+    )
+
+    # Check the defaults
+    assert np.all(
+        np.equal(np.isnan(layer_structure.lowest_canopy_filled), np.repeat(True, 9))
+    )
+    assert np.allclose(
+        layer_structure.get_indices("filled_canopy", as_bool=False), np.array([])
+    )
+
+    # Set an aggregate index
+    agg_roles = ["above", "filled_canopy", "surface"]
+    agg_index = frozenset(agg_roles)
+    agg_values = layer_structure.get_indices(agg_roles, as_bool=False)
+
+    assert agg_index in layer_structure._role_indices_int
+    assert np.all(np.equal(agg_values, np.array([0, 11])))
```

Comment:
Why does this not include any canopy indices?

Reply:
It's run _before_ `set_filled_canopy` is called and the default is that there is no canopy, so those layers are all empty. The intention is that this information can be set up before any models run, so we don't have to rely on any one model to populate `layer_heights` and then models can pop the the canopy into place.